### ISSUE TITLE
Unphysical dft energy from mcfun

### ISSUE THREAD

**Issue by xubwa** [Multi-collinear/MCfun#4]
Dear mcfun developers,
I recently tried to use mcfun in pyscf to run some really simple dft calculation, but the energy I got is very unphysical.  For a O2 molecule without any relativistic imposed, I used this input
```
%%time
from pyscf import dft, scf,gto
from pyscf import lib
lib.num_threads(4)
mol = gto.M(atom='''
O 0 0 0
O 3 0 0
''',
basis='631g',
ecp='631g',
verbose=4)
mf = dft.GKS(mol)
mf.kernel()
print(mf.collinear)
print(mf.energy_elec(), mf.energy_nuc(), mf.energy_tot())
```
First for a collinear test, and got a unconverged dft run, but the energy after 50 iteration is reasonable `SCF energy = -148.897389211094 after 50 cycles  <S^2> = 1.9999526  2S+1 = 2.9999684`
For a noncollinear calculation using mcfun with input 
```
mf = dft.GKS(mol)
#mf.with_soc = True
mf.collinear='mcol'
mf.kernel()
print(mf.energy_elec(), mf.energy_nuc(), mf.energy_tot())
```
I got 
```
converged SCF energy = -222.474529519497  <S^2> = 6.005868  2S+1 = 5.0023466
(-233.7636433524567, -4.909584147483386) 11.28911383296 -222.4745295194967
```
It seems to me the exchange correlation energy is incorrect. As I'm worrying that the newest pyscf might be  issue, so I made two clean environments and installed pyscf 2.1.0 and pyscf 2.2.0 respectively. Both yield the similar wrong numbers. The pyscf tests related to mcol functionals all failed. Do you guys have any idea where the problem might come from?

Thanks,
Xubo

**Comment by sunqm** [comment]
Does the energy look wrong since the first iteration? I'm not able to reproduce the problem. With the mcfun, I can get SCF like this
```
init E= -148.851698097305

WARN: HOMO -0.326683706919083 == LUMO -0.32655614813327

cycle= 1 E= -148.909360904386  delta_E= -0.0577  |g|= 0.0344  |ddm|= 1.09
  HOMO = -0.302555339812569  LUMO = -0.279830191425191
cycle= 2 E= -148.967796774703  delta_E= -0.0584  |g|= 0.0552  |ddm|= 1.39
  HOMO = -0.29064266920901  LUMO = -0.273227655870934
cycle= 3 E= -148.968446078174  delta_E= -0.000649  |g|= 0.0466  |ddm|= 0.0117

WARN: HOMO -0.246919221977438 == LUMO -0.246919221977436

cycle= 4 E= -148.984534119608  delta_E= -0.0161  |g|= 0.0219  |ddm|= 0.975
  HOMO = -0.247861558499269  LUMO = -0.245894308358949
cycle= 5 E= -148.984796802541  delta_E= -0.000263  |g|= 0.0142  |ddm|= 0.0124
...
```

**Comment by xubwa** [comment]
The energy was very wrong from the first iteration.

**Comment by Haskiy** [comment]
We make two tests. At the first test, we scan the potential curves of the triplet and the quintet oxygen molecule using UKS, respectively. It shows that the energy is lower for oxygen molecule with the bond length 3.0 Å at the quintet (-148.98498649 a.u.) than the triplet state (-148.93250119 a.u.).
At the second test, the GKS using multicollinear functional gives a calculation result, which is a quintet state, for the oxygen molecule with the bond length 3.0 Å. It is the same as the result given by UKS for the oxygen molecule at the quintet state.
`# GKS (mcol)
mol = gto.M(atom='''
O 0 0 0
O 3 0 0
''',
basis='631g',
ecp='631g',
verbose=4)
mf = dft.GKS(mol)
mf.collinear='mcol'
mf.kernel()
mf.e_tot, mf.mo_energy'

with output
**(-148.98498644404728,
 array([-18.75937473, -18.75934334, -18.70761544, -18.70758927,
         -0.89940989,  -0.89547329,  -0.783758  ,  -0.77872625,
         -0.38560538,  -0.38251006,  -0.35963423,  -0.35026766,
         -0.34691074,  -0.33851688,  -0.26300804,  -0.25167265,
         -0.24726211,  -0.24605111,  -0.24154718,  -0.23543278,
          0.64368988,   0.71129918,   0.71149325,   0.73292961,
          0.74855934,   0.77029058,   0.80019475,   0.80691445,
          0.80960164,   0.83747391,   0.84419503,   0.84701266,
          0.8802283 ,   0.92279114,   0.97255205,   1.04643068]))**

and 
`# UKS
mol = gto.M(atom='''
O 0 0 0
O 3 0 0
''',
basis='631g',
ecp='631g',
spin=4,
verbose=4)
mf = dft.UKS(mol)
mf.kernel()
mf.e_tot, mf.mo_energy '

with output
**(-148.9849864627382,
 array([[-18.75937473, -18.75934334,  -0.89940989,  -0.89547328,
          -0.38560538,  -0.38251005,  -0.35963423,  -0.35026766,
          -0.34691073,  -0.33851688,   0.64368988,   0.71129919,
           0.7329296 ,   0.74855936,   0.77029057,   0.80960164,
           0.84701266,   0.97255205],
        [-18.70761544, -18.70758926,  -0.78375799,  -0.77872624,
          -0.26300803,  -0.25167268,  -0.24726215,  -0.24605107,
          -0.24154714,  -0.23543278,   0.71149325,   0.80019471,
           0.80691451,   0.83747388,   0.84419508,   0.8802283 ,
           0.92279115,   1.04643068]]))**.

However, we didn't add the soc term in our tests, which will not change our conclusion. And we couldn't reproduce your result.

**Comment by xubwa** [comment]
Someone help prepared a docker file which includes both mcfun and pyscf, the wrong energy can be reproduced there, could you try run that? Perhaps there's something on the release version has some discrepancy with the development version you guys have. The docker image can be pulled from `docker pull hblack35/pyscf:mcfun`, or simply use this dockerfile
```
FROM python:3.7-slim

LABEL maintainer="Qiming Sun <osirpt.sun@gmail.com>"

RUN pip install --no-cache-dir pyscf mcfun
```

**Comment by sunqm** [comment]
Problem is fixed in the latest release v0.2.2
